modules/geochemistry/src/actions/AddGeochemistrySolverAction.C

```diff
@@ -0,0 +1,344 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AddGeochemistrySolverAction.h"
+#include "GeochemicalModelDefinition.h"
+
+registerMooseAction("GeochemistryApp", AddGeochemistrySolverAction, "add_output");
+registerMooseAction("GeochemistryApp",
+                    AddGeochemistrySolverAction,
+                    "add_geochemistry_molality_aux");
+
+InputParameters
+AddGeochemistrySolverAction::validParams()
+{
+  InputParameters params = Action::validParams();
+  params.addRequiredParam<UserObjectName>("model_definition",
+                                          "The name of the GeochemicalModelDefinition user object "
+                                          "(you must create this UserObject yourself)");
+  params.addParam<UserObjectName>(
```

Comment:
FYI - as you seem to be growing these large lists of params in your actions that are duplicates of the objects they create, I have a suggestion. Take, for example, `GeochemistryConsoleOutput`.

You can create an additional param static function, say `static InputParameters GeochemistryConsoleOutput::sharedParams()`. Here is where you would put the params that you want to appear both in the console output object and the action that creates it. Then, you would have
```
InputParameters
GeochemistryConsoleOutput::validParams()
{
  InputParameters params = Output::validParams();
  params += GeochemistryConsoleOutput::sharedParams();
}
```
This would then allow you to do something similar here, in which you can just add all of the shared params from another object without having to duplicate them.

Reply:
Oh cool, i didn't know about that.  Thanks - i'll address that today

Reply:
Done the sharedParams idea, thanks @loganharbour - got rid of 262 lines of duplicated code!

Reply:
Awesome! Looks great!

Reply:
Thanks @loganharbour  :-)